keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -86,42 +120,133 @@ become possible.
 
 ### Goals
 
-The goal is to enable a union or "oneof" semantics in Kubernetes types, both for
-in-tree types and for CRDs.
+Provide a simple mechanism for API authors to label fields of a resource as
+members of a oneOf, in order to receive standardized:
+
+* Validation - ensuring only one member field is set (or at most one if
+  desired).
+* Normalization - ensuring the API server can modify the fields of the oneOf to
+  best match the intent of the client, despite the client potentially having
+  incomplete information
 
 ### Non-Goals
 
-We're not planning to use this KEP to release the feature, but mostly as a way
-to document what we're doing.
+The focus of this KEP is on providing unified validation and normalization of
+new union types. Migrating existing unions away from their bespoke validation
+logic (e.g validation functions), is an explicit non-goal and will be pursued in
+a separate KEP or later release.
```

Comment:
I think we should do at least one of each type of union, as a check that our proposal actually works. I agree completeness is not a goal at this point.

Reply:
I'd want to see it in alpha, because otherwise I'm not sure we're building the right thing. Building something with a difficult/impossible migration path is also a failure mode.

Reply:
> Building something with a difficult/impossible migration path is also a failure mode.

While migration would be ideal, providing something that work for future types still provides significant value and would allow us to avoid our past mistakes and limitations.

Reply:
see updated note in graduation criteria